File: src/data/updateForms.py
# -*- coding: utf-8 -*-
import csv
import mysql.connector
from mysql.connector import Error

def read_tsv(file_path):
    tsv_data = {}
    with open(file_path, 'r', newline='', encoding='utf-8') as tsvfile:
        reader = csv.DictReader(tsvfile, delimiter='\t')
        for row in reader:
            try:
                moph_code = row['MoPHCode'].strip()  # Clean TSV codes
                form = row['Form'].strip()
                tsv_data[moph_code] = form
            except KeyError:
                print(f"Skipping invalid row: {row}")
    return tsv_data

def get_db_connection():
    try:
        return mysql.connector.connect(
            host='localhost',
            user='ommal_ahmad',
            password='fISfGr^8q!_gUPMY',
            database='ommal_medapiv2'
        )
    except Error as e:
        print(f"Error: {e}")
        return None

def main():
    tsv_data = read_tsv('RPFD.tsv')
    
    conn = get_db_connection()
    if conn is None:
        print("Failed to connect to the database.")
        return

    try:
        conn.autocommit = False
        cursor = conn.cursor(dictionary=True)

        # Fetch and clean database codes
        cursor.execute("SELECT MoPHCode, Form FROM drug")
        current_data = {
            row['MoPHCode'].strip(): row['Form']  # Clean DB codes
            for row in cursor.fetchall()
        }

        # Calculate changes
        update_list = []
        delete_list = []

        # Process TSV codes (5578 entries)
        for code, new_form in tsv_data.items():
            current_form = current_data.get(code)
            if current_form != new_form:
                update_list.append((new_form, code))

        # Process codes in DB but not in TSV
        delete_list = [
            code for code, form in current_data.items()
            if code not in tsv_data and form is not None
        ]

        # Show preview
        print("\n===== Changes Preview =====")
        print(f"Updates to perform: {len(update_list)}")
        if update_list:
            print("Sample updates:")
            for update in update_list[:3]:
                print(f"Code {update[1]}: '{current_data.get(update[1])}' → '{update[0]}'")

        print(f"\nForms to clear: {len(delete_list)}")
        if delete_list:
            print("Sample deletions:")
            for code in delete_list[:3]:
                print(f"Code {code}: '{current_data[code]}' → NULL")

        # Confirmation
        confirm = input("\nDo you want to commit these changes? (yes/no): ").lower()
        if confirm != 'yes':
            conn.rollback()
            print("Transaction rolled back")
            return

        # Execute updates
        if update_list:
            update_query = "UPDATE drug SET Form = %s WHERE MoPHCode = %s"
            cursor.executemany(update_query, update_list)

        # Execute deletions
        if delete_list:
            delete_query = "UPDATE drug SET Form = NULL WHERE MoPHCode = %s"
            cursor.executemany(delete_query, [(code,) for code in delete_list])

        conn.commit()
        print(f"\nSuccessfully committed changes:")
        print(f"- {len(update_list)} updates")
        print(f"- {len(delete_list)} deletions")

    except Error as e:
        print(f"\nDatabase error: {e}")
        conn.rollback()
    finally:
        if conn.is_connected():
            cursor.close()
            conn.close()

if __name__ == "__main__":
    main()

File: src/data/updateForms.js
const fs = require('fs');
const csv = require('csv-parser');
const mysql = require('mysql2/promise');

async function readTsv(filePath) {
    const tsvData = {};
    return new Promise((resolve, reject) => {
        fs.createReadStream(filePath)
            .pipe(csv({ separator: '\t' }))
            .on('data', (row) => {
                try {
                    const mophCode = row['MoPHCode'].trim();
                    const form = row['Form'].trim();
                    tsvData[mophCode] = form;
                } catch (error) {
                    console.log(`Skipping invalid row: ${JSON.stringify(row)}`);
                }
            })
            .on('end', () => {
                resolve(tsvData);
            })
            .on('error', (error) => {
                reject(error);
            });
    });
}

async function getDbConnection() {
    try {
        return await mysql.createConnection({
            host: 'localhost',
            user: 'ommal_ahmad',
            password: 'fISfGr^8q!_gUPMY',
            database: 'ommal_medapiv2'
        });
    } catch (error) {
        console.error(`Error: ${error.message}`);
        return null;
    }
}

async function main() {
    const tsvData = await readTsv('RPFD.tsv');
    const conn = await getDbConnection();
    if (!conn) {
        console.log("Failed to connect to the database.");
        return;
    }

    try {
        await conn.beginTransaction();

        // Fetch current data from the database
        const [rows] = await conn.execute("SELECT MoPHCode, Form FROM drug");
        const currentData = {};
        rows.forEach(row => {
            currentData[row.MoPHCode.trim()] = row.Form;
        });

        // Calculate changes
        const updateList = [];
        for (const [code, newForm] of Object.entries(tsvData)) {
            const currentForm = currentData[code];
            if (currentForm !== newForm) {
                updateList.push([newForm, code]);
            }
        }

        // Show preview
        console.log("\n===== Changes Preview =====");
        console.log(`Updates to perform: ${updateList.length}`);
        if (updateList.length > 0) {
            console.log("Sample updates:");
            updateList.slice(0, 3).forEach(update => {
                console.log(`Code ${update[1]}: '${currentData[update[1]]}' → '${update[0]}'`);
            });
        }

        // Confirmation
        const confirm = require('readline-sync').question("\nDo you want to commit these changes? (yes/no): ").toLowerCase();
        if (confirm !== 'yes') {
            await conn.rollback();
            console.log("Transaction rolled back");
            return;
        }

        // Execute updates
        if (updateList.length > 0) {
            const updateQuery = "UPDATE drug SET Form = ? WHERE MoPHCode = ?";
            await conn.query(updateQuery, [updateList]);
        }

        await conn.commit();
        console.log(`Successfully committed ${updateList.length} changes`);

    } catch (error) {
        console.error(`Database error: ${error.message}`);
        await conn.rollback();
    } finally {
        await conn.end();
    }
}

main().catch(error => console.error(error));